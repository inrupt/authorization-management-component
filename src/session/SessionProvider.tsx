// MIT License
//
// Copyright Inrupt Inc.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.
//

import type { ReactElement, ReactNode } from "react";
import {
  createContext,
  useCallback,
  useEffect,
  useMemo,
  useState,
} from "react";

import type { Session } from "@inrupt/solid-client-authn-browser";
import {
  fetch,
  getDefaultSession,
  login,
  logout,
} from "@inrupt/solid-client-authn-browser";

import { useRouter } from "next/router";
import { getPodUrlAllFrom, getProfileAll } from "@inrupt/solid-client";
import { getAccessApiEndpoint } from "@inrupt/solid-client-access-grants";
import { getVerifiableCredentialApiConfiguration } from "@inrupt/solid-client-vc";
import type { VerifiableCredentialApiConfiguration } from "@inrupt/solid-client-vc/dist/common/common";
import {
  WebIdNotAvailable,
  DiscoveryNotAvailable,
  AccessGrantsNotSupported,
} from "../../src/errors/customErrors";
import { ProfileModal } from "../components/ProfileModal/ProfileModal";
import { ErrorModal } from "../components/ErrorModal/ErrorModal";

type AuthState =
  | "waiting"
  | "authenticating"
  | "unauthenticated"
  | "authenticated"
  | "pre-authenticating";

export interface ISessionContext {
  login: typeof login;
  logout: typeof logout;
  session: Session;
  fetch: typeof fetch;
  sessionRequestInProgress: boolean;
  state: AuthState;
  accessEndpoint?: string;
  endpointConfiguration?: VerifiableCredentialApiConfiguration;
}

export const createContextBase: () => ISessionContext = () => ({
  login,
  logout,
  fetch,
  session: getDefaultSession(),
  sessionRequestInProgress: true,
  accessEndpoint: undefined,
  state: "waiting",
  endpointConfiguration: undefined,
});

export const SessionContext =
  createContext<ISessionContext>(createContextBase());

/* eslint react/require-default-props: 0 */
export interface ISessionProvider {
  children: ReactNode;
  onError?: (_error: Error) => void;
  restorePreviousSession?: boolean;
}

/**
 * @param session
 * @param callback The callback should mutate react state to trigger re-rendering.
 */
const discoverAccessEndpoint = async (
  session: Session,
  callback: (endpoint: string) => void,
): Promise<void> => {
  if (typeof session.info.webId !== "string") {
    return;
  }
  let profiles;
  try {
    profiles = await getProfileAll(session.info.webId, {
      fetch: session.fetch,
    });
  } catch (e: unknown) {
    throw new WebIdNotAvailable(
      `An error occured when looking up WebID ${session.info.webId}: ${e}`,
    );
  }
  const pods = getPodUrlAllFrom(profiles, session.info.webId);
  if (pods.length < 1) {
    throw new Error(
      `No Pods could be discovered from WebID ${session.info.webId}`,
    );
  }
  // FIXME arbitrarily pick the first Pod.
  const [podRoot] = pods;
  let endpoint;
  try {
    endpoint = await getAccessApiEndpoint(podRoot);
  } catch (e: unknown) {
    throw new DiscoveryNotAvailable(
      `Access endpoint discovery for Pod ${podRoot} failed: ${e}`,
    );
  }
  callback(endpoint);
};

const HintModal = ({
  error,
  modalStatus,
  modalControl,
}: {
  error?: Error;
  modalStatus: boolean;
  modalControl: (status: boolean) => void;
}) => {
  if (error === undefined) {
    return <></>;
  }
  if (error instanceof WebIdNotAvailable) {
    return (
      <ProfileModal
        isOpen={modalStatus}
        onClose={() => {
          modalControl(false);
        }}
      />
    );
  }
  if (error instanceof DiscoveryNotAvailable) {
    return (
      <ErrorModal
        err={
          new AccessGrantsNotSupported(
            "Your Pod Provider doesn't appear to support Access Grants. Please retry using a compliant Pod Provider.",
            {
              cause: error,
            },
          )
        }
        isOpen={modalStatus}
        onClose={() => {
          modalControl(false);
        }}
      />
    );
  }
  return (
    <ErrorModal
      err={error}
      isOpen={modalStatus}
      onClose={() => {
        modalControl(false);
      }}
    />
  );
};

/**
 * Used to provide session data to child components through context.
 */
export default function SessionProvider({
  children,
  onError,
  restorePreviousSession,
}: ISessionProvider): ReactElement {
  const { replace, isReady } = useRouter();
  const session = useMemo(getDefaultSession, []);

  const [state, setState] = useState<AuthState>(
    session.info.isLoggedIn ? "authenticated" : "waiting",
  );

  const [accessEndpoint, setAccessEndpoint] = useState<string>();
  const [accessEndpointConfig, setAccessEndpointConfig] =
    useState<VerifiableCredentialApiConfiguration>();
  const [modalOpen, setModalOpen] = useState<boolean>(true);
  const [errorState, setErrorState] = useState<Error>();

  const isAuthenticated = state === "authenticated";
  const isUnAuthenticated = state === "unauthenticated";

  useEffect(() => {
    if (isReady) {
      if (state === "waiting") {
        setState("authenticating");
      } else if (state === "authenticating") {
        session
          .handleIncomingRedirect({
            url: window.location.href,
            restorePreviousSession,
          })
          .then((res) => {
            setState(res?.isLoggedIn ? "authenticated" : "unauthenticated");
          })
          .catch((error) => {
            if (onError) {
              onError(error as Error);
            } else {
              throw error;
            }
            setState(
              session.info?.isLoggedIn ? "authenticated" : "unauthenticated",
            );
          });
      }
    }
  }, [state, isReady, onError, replace, session, restorePreviousSession]);

  useEffect(() => {
    session.events.on("logout", () => {
      if (isAuthenticated || isUnAuthenticated) {
        setState("waiting");
      }
    });
  }, []);

  useEffect(() => {
    if (state === "authenticated") {
      discoverAccessEndpoint(session, (endpoint: string) => {
        setAccessEndpoint(endpoint);
      }).catch((e) => {
        if (e instanceof Error) {
          setErrorState(e);
        } else {
          console.error(`An error occured: ${e}`);
        }
      });
    }
  }, [state]);

  useEffect(() => {
    if (typeof accessEndpoint === "string") {
      getVerifiableCredentialApiConfiguration(accessEndpoint).then(
        setAccessEndpointConfig,
        (error: unknown) => {
          if (error instanceof Error) {
            setErrorState(error);
          } else {
            console.error(`An error occured: ${error}`);
          }
        },
      );
    }
  }, [accessEndpoint]);

  const loginCallback = useCallback(
    async (options: Parameters<Session["login"]>[0]) => {
      if (isUnAuthenticated) {
        setState("pre-authenticating");
        try {
          await session.login(options);
        } catch (error) {
          if (onError) {
            onError(error as Error);
          } else {
            throw error;
          }
        } finally {
          setState("unauthenticated");
        }
      }
    },
    [isUnAuthenticated, onError, session],
  );

  const logoutCallback = useCallback(
    async (options: Parameters<Session["logout"]>[0]) => {
      if (isAuthenticated) {
        try {
          await session.logout(options);
          setState("unauthenticated");
        } catch (error) {
          if (onError) {
            onError(error as Error);
          } else {
            throw error;
          }
        }
      }
    },
    [isAuthenticated, onError, session],
  );

  const value = useMemo<ISessionContext>(
    () => ({
      login: loginCallback,
      logout: logoutCallback,
      fetch: session.fetch,
      session,
      sessionRequestInProgress:
        state === "waiting" || state === "authenticating",
      state,
      accessEndpoint,
      endpointConfiguration: accessEndpointConfig,
    }),
    [
      loginCallback,
      logoutCallback,
      state,
      session,
      accessEndpoint,
      accessEndpointConfig,
    ],
  );

  return (
    <SessionContext.Provider value={value}>
      <>
        <HintModal
          modalControl={setModalOpen}
          modalStatus={modalOpen}
          error={errorState}
        ></HintModal>
        {children}
      </>
    </SessionContext.Provider>
  );
}
