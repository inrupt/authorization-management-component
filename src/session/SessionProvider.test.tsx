// MIT License
//
// Copyright Inrupt Inc.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.
//

import * as React from "react";
import { act, render, waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
// There appears to be a TS issue with jest globals, but testing-library/react expects them to be present.
// eslint-disable @typescript-eslint/ban-ts-comment
// import { jest } from "@jest/globals";

import type { NextRouter, useRouter } from "next/router";
import type SolidClient from "@inrupt/solid-client";
import type { getProfileAll, getPodUrlAllFrom } from "@inrupt/solid-client";
import type SolidClientAuthnBrowser from "@inrupt/solid-client-authn-browser";
import type { Session } from "@inrupt/solid-client-authn-browser";
import type { EventEmitter } from "stream";
import SessionProvider, { SessionContext } from "./SessionProvider";

jest.mock("@inrupt/solid-client-authn-browser");
jest.mock("@inrupt/solid-client", () => {
  const actualSolidClient = jest.requireActual<typeof SolidClient>(
    "@inrupt/solid-client",
  );
  return {
    // @ts-expect-error Using jest globals causes issues with jest signatures.
    getProfileAll: jest.fn<typeof getProfileAll>().mockResolvedValue({
      webIdProfile: actualSolidClient.mockSolidDatasetFrom(
        "https://example.org/id",
      ),
      altProfileAll: [],
    }),
    // @ts-expect-error Using jest globals causes issues with jest signatures.
    getPodUrlAllFrom: jest.fn<typeof getPodUrlAllFrom>().mockReturnValue([]),
  };
});
jest.mock("next/router", () => ({
  useRouter: jest.fn().mockReturnValue({
    replace: jest.fn(),
    isReady: true,
  }),
}));

function ChildComponent(): React.ReactElement {
  const {
    session,
    sessionRequestInProgress,
    login: sessionLogin,
    logout: sessionLogout,
    state,
  } = React.useContext(SessionContext);

  return (
    <div>
      {sessionRequestInProgress && (
        <div data-testid="sessionRequestInProgress">
          sessionRequestInProgress
        </div>
      )}
      <div data-testid="state">{state}</div>
      <div data-testid="session">{JSON.stringify(session)}</div>
      <button type="button" onClick={async () => sessionLogin({})}>
        Login
      </button>
      <button
        type="button"
        onClick={async () => sessionLogout()}
        data-testid="logout"
      >
        Logout
      </button>
    </div>
  );
}

describe("Testing SessionContext", () => {
  let onError: jest.Mock;

  // @ts-expect-error Using jest globals causes issues with jest signatures.
  const handleIncomingRedirect = jest.fn<Session["handleIncomingRedirect"]>();
  let user: ReturnType<typeof userEvent.setup>;
  let session: Session;
  let documentBody: ReturnType<typeof render>;

  beforeEach(() => {
    session = {
      info: {
        isLoggedIn: false,
        webId: "https://fakeurl.com/me",
        sessionId: "abc123",
      },
      on: jest.fn(),
      handleIncomingRedirect,
      events: {
        on: jest.fn(),
      } as Partial<EventEmitter> as EventEmitter,
      login: jest.fn(),
      logout: jest.fn(),
    } as Partial<Session> as Session;
    handleIncomingRedirect.mockResolvedValue(session.info);

    jest.requireMock<typeof SolidClientAuthnBrowser>(
      "@inrupt/solid-client-authn-browser",
    ).getDefaultSession = jest
      // @ts-expect-error Using jest globals causes issues with jest signatures.
      .fn<typeof SolidClientAuthnBrowser.getDefaultSession>()
      .mockReturnValue(session);
    onError = jest.fn();
    user = userEvent.setup();
  });

  it("calls onError if handleIncomingRedirect fails", async () => {
    handleIncomingRedirect.mockRejectedValueOnce(undefined);
    jest.requireMock<{
      useRouter: typeof useRouter;
      // @ts-expect-error Using jest globals causes issues with jest signatures.
    }>("next/router").useRouter = jest.fn<typeof useRouter>().mockReturnValue({
      replace: jest.fn(),
      isReady: true,
    } as unknown as NextRouter);
    render(
      <SessionProvider onError={onError}>
        <ChildComponent />
      </SessionProvider>,
    );

    await waitFor(() => {
      expect(onError).toHaveBeenCalled();
    });
  });

  it("uses the logout functions from session", async () => {
    session.info.isLoggedIn = true;
    act(() => {
      documentBody = render(
        <SessionProvider>
          <ChildComponent />
        </SessionProvider>,
      );
    });

    await waitFor(() => {
      expect(handleIncomingRedirect).not.toHaveBeenCalled();
    });

    const { getByText } = documentBody;

    expect(session.login).not.toHaveBeenCalled();
    expect(session.logout).not.toHaveBeenCalled();

    await user.click(getByText("Login"));
    expect(session.login).toHaveBeenCalledTimes(0);
    await user.click(getByText("Logout"));
    expect(session.logout).toHaveBeenCalledTimes(1);
  });

  describe("Resolving handleIncomingRedirect", () => {
    beforeEach(() => {
      handleIncomingRedirect.mockResolvedValueOnce(
        Promise.resolve(session.info),
      );
    });

    it("matches snapshot", async () => {
      act(() => {
        documentBody = render(
          <SessionProvider>
            <ChildComponent />
          </SessionProvider>,
        );
      });

      await waitFor(() => {
        expect(session.handleIncomingRedirect).toHaveBeenCalled();
      });

      const { getByText } = documentBody;
      await waitFor(() => {
        expect(getByText("unauthenticated")).toBeInTheDocument();
      });

      const { baseElement } = documentBody;
      expect(baseElement).toMatchSnapshot();
    });

    it("uses the login functions from session", async () => {
      act(() => {
        documentBody = render(
          <SessionProvider>
            <ChildComponent />
          </SessionProvider>,
        );
      });

      await waitFor(() => {
        expect(handleIncomingRedirect).toHaveBeenCalled();
      });

      const { getByText } = documentBody;
      await waitFor(() => {
        expect(getByText("unauthenticated")).toBeInTheDocument();
      });

      expect(session.login).not.toHaveBeenCalled();
      expect(session.logout).not.toHaveBeenCalled();

      await user.click(getByText("Login"));
      expect(session.login).toHaveBeenCalledTimes(1);
      await user.click(getByText("Logout"));
      expect(session.logout).toHaveBeenCalledTimes(0);
    });
  });

  it.each(["Login", "Logout"])(
    "calls onError if there is an error during %s",
    async (selector) => {
      (
        session[selector.toLowerCase() as "login" | "logout"] as jest.Mock
      ).mockRejectedValueOnce(null);

      session.info.isLoggedIn = selector !== "Login";

      session.handleIncomingRedirect = jest.fn(async () => session.info);

      const { getByText } = render(
        <SessionProvider onError={onError}>
          <ChildComponent />
        </SessionProvider>,
      );

      await waitFor(() => {
        expect(session.handleIncomingRedirect).toHaveBeenCalledTimes(
          selector === "Login" ? 1 : 0,
        );
      });

      await user.click(getByText(selector));

      expect(onError).toHaveBeenCalledTimes(1);
    },
  );
});
