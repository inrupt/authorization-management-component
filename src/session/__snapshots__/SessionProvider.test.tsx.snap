// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing SessionContext Resolving handleIncomingRedirect matches snapshot 1`] = `
<body
  class=""
  style=""
>
  <div>
    <div>
      <div
        data-testid="state"
      >
        unauthenticated
      </div>
      <div
        data-testid="session"
      >
        {"info":{"isLoggedIn":false,"webId":"https://fakeurl.com/me","sessionId":"abc123"},"events":{}}
      </div>
      <button
        type="button"
      >
        Login
      </button>
      <button
        data-testid="logout"
        type="button"
      >
        Logout
      </button>
    </div>
  </div>
</body>
`;
